### ISSUE TITLE
celluloid-extras

### ISSUE THREAD

**Issue by digitalextremist** [celluloid/celluloid#549]
Going to start a gem called `celluloid-extras` for #371 and #536, preparing for more and encouraging a `contrib` style gem. Why I'd say this gem exist versus individual gems for each is that there is a lot of duplicated work in providing access to the `Celluloid` test suite and `Celluloid` itself, which can be worked around once and provided to all adopted sub-libraries. Then all can also be updated at once when `Celluloid` advances.  Each involved author will be attributed, and I will coordinate licensing with each.

**Comment by digitalextremist** [comment]
I'm actually torn on whether these are one gem or separate. If this were an omnibus style repository it'd be easier, but maintaining separate gems seems non-awesome... and I'd like to be able to update gems whenever the root API is changed, until and perhaps after `1.0`

@tarcieri, so far I have:

- `celluloid-pool`
- `celluloid-chain`
- `celluloid-coordinator`
- `celluloid-task-pooled-fiber`
- `celluloid-future`
- `celluloid-promise`

**Comment by e2** [comment]
> but maintaining separate gems seems non-awesome

If I can chime in, I'd say to aim for separation as much as possible, but no sooner than the core API is polished.

Early separation will kill API improvement efforts (arch-level refactoring), because it would mean introducing the baggage of compatibility.

But the biggest benefit of separation is *no longer* having to maintain functionality people don't care enough about. Second on the list is: ironically, it actually helps define a better core API.

As for specs, they shouldn't be shared at all. E.g. Celluloid::IO shouldn't use actor specs - it should test only what it "adds on" to Celluloid. It's actually Celluloid that should have an "API" for extending it.

E.g. There should be specs so I can do something like:
```
ruby
class MyActor
  include MyCelluloid
  # (...)
end
```
where e.g. 'MyCelluloid' could include e.g. promises and evented mailboxes and/or Celluloid::IO. Otherwise duplicating tests means both gems are unnecessarily sharing responsibilities.

It may seem complex, but I think "extensibility" is the biggest challenge when trying to extract things outside from Celluloid. Partially because greater extensibility may hurt performance a lot as well.

**Comment by digitalextremist** [comment]
I welcome your thoughts @e2.

I agree about sequence - hardening the API before dispersing gems.

I agree about not sharing specs. But in the case of `Task::PooledFiber` I needed `behaves as Celluloid task` so I ended up just duplicating the spec for that ( and died a little bit inside ).

I like the idea of rolling your own `Celluloid` includable object.

An extensions API is starting to shape up. We're finding areas where alternatives ought to exist for a certain approach, and where additional functionality could be naturally added with enough frequency to justify additional thought going into extensibility. But you are also right about performance hits, I keep wanting to include "drop in" modifications with official support ( to remove for example my own overriding of `Celluloid` internals ( which inevitably ends up with cascading refactor work ) ) but even just a test for `nil` and replacing it with a default is too much overhead for me usually.

So far I'd say I'm personally at 66% to 75% "for" dismantling the gems and keeping them individual.

**Comment by e2** [comment]
Thanks.

TL;DR - suggestion: "Celluloid compliance gem" for plugin/extension authors.

Specs have "version dependencies" too - they depend on a specific "internal-API-compatible" snapshot of a codebase, so sharing them between gems doesn't make sense. (They often depend too much on the "internal API").

Plugins though (like Celluloid::IO, custom tasks, custom mailboxes) ... they need their own specs that check if they "comply" to Celluloids *public* API. But nothing more, really.

Just saying there's a clear distinction between "unit tests" and "compliance tests".

Since Celluloid doesn't have a notion of "extentions" or "plugins", it's easy to forget that "compliance tests" are either buried somewhere within the unit test (and need to be extracted to a separate gem), or need to be written.

It may seem like overhead, but I'd argue is saves more time and brings core-devs a lot more freedom and flexibility:
 
1. Anyone writing a "celluloid plugin", goes to the "celluloid-compat-plugin" (or whatever) gem's homepage, and follows instructions - along with setting up "compliance" tests. There'd be plugin-API-specific README, wiki, issues, examples, etc.
2. Any requests in core for new stuff could be handled with: "write an extension" and redirect to the compliance gem's page
3. since custom plugin will depend only on "celluoid-compat-plugin" (and not on Celluloid), ALL of Celluoid can be stubbed/mocked in the specs (!!!)
4. Celluloid core-devs can handle extension/plugin-related issues in one place, adding more "compliance" tests as needed (e.g. freezing objects to make sure they don't pass state where they shouldn't, testing for sleep() calls in initialize, checking if finalizers survive where necessary, etc.)
4. Celluloid core-devs can "soften" compatibility issues by moving all bad "hacks" and "workarounds" into this "compliance" gem, keeping core "nice", while plugin authors have a safe "sandbox"
5. New features can first be added here instead of core (unless impossible) - without affecting core releases
6. Instead of "adding stuff to Celluloid", plugin authors could just request additional APIs and extensions in the 'compliance" gem (keeping issue+PR handling in "core" Celluloid lean and clean).
7. Performance issues can be "outsourced" somewhat - either people stick to a "bare-metal" "core" Celluloid, or they use extensions "implying" that they don't mind the extensive overhead that comes with the "compliance" gem (which could have non-optimal but dev-friendly alternatives to Celluloid's core classes, e.g. even "slow-but-fully-traceable" SyncCall)
8. Plugins that people loose interest in become "dead" and it would not be a baggage "dragged around" "just in case someone's relying on it". If a plugin is complex but needed, people will update it to keep it "compliant" (instead of the core team stressing about compatibility for "non-existing users").
9. You can add version-specific warnings/incompatibilities in this "compliance" gem, redirect old API usage to new API usage, switch implementations/workarounds based on Celluloid version
10. Core-devs have the full freedom to decide which duplication to remove, or which to keep/create (for performance). E.g. SyncCall has to be lean in core, but if "celluloid-plugin-compat" is included, a "plugin-friendly" alternative is used (extra checks, before/after filters, "safe mode" with freezing, tracing) - whether both would have shared logic or not - that's a decision in the "compliance" gem, not in core.
11. This also means plugin/extension authors could have a place for their own "community" (a separation of non-core discussions from core discussions).

So before splitting, I'd consider 3 APIs:
1. Public API (for production and SemVer) - minimal enough just so "people can somehow achieve what they want without monkey-patching". It's only in the docs if it's a Public API (and vice verse). E.g. I'd say `SyncCall` is *not* public API (and shouldn't be in the YARD docs).
2. Internal Celluloid API - which can change at any moment. No mention in the docs anywhere, which means core devs can refactor/modify at any moment without worrying about compatibility.
3. Plugin API defined in a separate gem (which may or may not depend on Celluloid internal API) - a platform to "organize" and "negotiate" features/changes for extensions - without affecting work on core or performance.

Just my 2 cents... given I see too many "plugins" heavily "integrated" inside core, while I'm constantly "manually" adding my own features (like "tracing" to work things out, or removing notification and incident reporting for "less things happening").

I think it seems natural to "eat your own dogfood" and first develop a "compliance gem" and then "follow it's instructions" while extracting things out of "core".

(Back to debugging that Probe spec race condition in JRuby...)

**Comment by tarcieri** [comment]
I'd just focus on extracting pools for now...

**Comment by e2** [comment]
I agree - pools don't even belong in core.

**Comment by digitalextremist** [comment]
@e2 that was $2! :) Food for thought. Will come back to that when we implement/encourage `celluloid-*` gems.

@tarcieri pulling pools out first is the plan proto-ish; thats separate of setting a precedent to me. I was more generally wondering out-loud if that is `celluloid-pool` and all others are `celluloid-*` or, if we have a catch-all `celluloid-extras`, but it seems we'll be doing individual gems.

**Comment by digitalextremist** [comment]
This was created and contains `Rehasher` and a concurrent `Hash` implementation, whereas all the other mentioned gems were created first, and stand alone.
